File: algorithms/sorting/bubble_sort/bubble_sort.py
from typing import List, TypeVar

T = TypeVar('T')


def bubble_sort(arr: List[T]) -> None:
    """
    >>> arr = [3,2,1,4,5]
    >>> bubble_sort(arr)
    >>> print(arr)
    [1, 2, 3, 4, 5]
    """
    n = len(arr) - 1
    for i in range(n):
        swapped = False
        for j in range(n - i):
            if arr[j] > arr[j+1]:
                arr[j+1], arr[j] = arr[j], arr[j+1]
                swapped = True

        if not swapped:
            break


if __name__ == '__main__':
    import doctest
    doctest.testmod()

File: algorithms/sorting/bubble_sort/bubbleSort.ts
import { swapArrayIndexes } from '@utils';

function bubbleSort<T>(arr: T[]): void {
    const len = arr.length;
    if (len <= 1) {
        return;
    }
    // start from the beginning
    for (let i = 0; i < len; i++) {
        // we need this variable to be able to tell if array is sorted or not
        let swapped = false;
        // inner loop will be till the end of array minus `i`, since after every
        // iteration last element will be sorted
        for (let j = 0; j < len - i; j++) {
            if (arr[j] > arr[j + 1]) {
                // Every element that is larger gets bubbled to the end of the array
                swapArrayIndexes(arr, j, j + 1);
                swapped = true;
            }
        }
        // if we didn't swap during the iteration that means that array is sorted
        // and there is no need to iterate over the array, we break the loop
        if (!swapped) {
            break;
        }
    }
}

export default bubbleSort;
